src/Language/PureScript/Sugar/LetPattern.hs

```diff
@@ -0,0 +1,50 @@
+-- |
+-- This module implements the desugaring pass which replaces patterns in let-in
+-- expressions with appropriate case expressions.
+--
+module Language.PureScript.Sugar.LetPattern (desugarLetPatternModule) where
+
+import Prelude.Compat
+
+import Control.Monad.Error.Class (MonadError(..))
+
+import Language.PureScript.AST
+import Language.PureScript.Errors
+
+-- |
+-- Replace every @BoundValueDeclaration@ in @Let@ expressions with @Case@
+-- expressions.
+--
+desugarLetPatternModule :: forall m. (MonadError MultipleErrors m) => Module -> m Module
+desugarLetPatternModule (Module ss coms mn ds exts) = Module ss coms mn <$> parU ds desugarLetPattern <*> pure exts
+
+-- |
+-- Desugar a single let expression
+--
+desugarLetPattern :: forall m. (MonadError MultipleErrors m) => Declaration -> m Declaration
```

Comment:
`rethrowWithPosition` needs the constraint, but you don't need `rethrowWithPosition`, since nothing will be thrown 😄